### ISSUE TITLE
ddet and drand not very adversarial?

### ISSUE THREAD

**Issue by vikramnitin9** [MadryLab/constructed-datasets#11]
If my understanding is correct, a Resnet50 model trained on the original CIFAR dataset should perform very well on the ddet and drand datasets, because these datasets are constructed by adversarially perturbing every image in CIFAR with respect to a (trained) Resnet50 model. If these adversarial attacks are "strong enough", one would expect the reference model to get an accuracy of close to 100% on ddet and drand.

However in the below example, I train a Resnet50 model on CIFAR and evaluate it on ddet and drand. I get accuracies of 38% and 42% respectively, which are a long way from the expected near 100% accuracy. Further, the model has an accuracy of 57% on the "reverse-shifted" ddet dataset, which suggests that it still prefers the original labels to the adversarial labels.

I've used your robustness library to make it easier for you to verify that my implementation has no errors. I'd be grateful for some help in trying to figure out what I'm doing wrong here.

```python
# A minimal example to demonstrate that the MadryLab ddet and drand datasets
# are not very adversarial to a model trained on CIFAR

# Create and populate these directories
cifar_root          = 'data/cifar-10-python'
release_data_root   = 'data/release_datasets'
save_dir            = 'saved_models/madry_cifar_train'

import torch
from torchvision import transforms
import os

from robustness.train       import train_model, eval_model
from robustness.datasets    import CIFAR
from robustness.model_utils import make_and_restore_model

class AttrDict(dict):
    def __init__(self, *args, **kwargs):
        super(AttrDict, self).__init__(*args, **kwargs)
        self.__dict__ = self

class ReleaseDatasets(torch.utils.data.Dataset):

    def __init__(self, root='data/release_datasets/ddet_CIFAR', transform=transforms.ToTensor(), target_transform=lambda x:x, **kwargs):

        self.data   = torch.cat(torch.load(os.path.join(root, 'CIFAR_ims')))
        self.labels = torch.cat(torch.load(os.path.join(root, 'CIFAR_lab')))

        self.transform          = transforms.Compose([transforms.ToPILImage(), transform])
        self.target_transform   = target_transform

        super(ReleaseDatasets, self).__init__()

    def __len__(self):
        return self.data.shape[0]

    def __getitem__(self, idx):
        return self.transform(self.data[idx]), self.target_transform(self.labels[idx])

class ShiftedDDet(ReleaseDatasets):

    def __init__(self, **kwargs):
        super(ShiftedDDet, self).__init__(target_transform = lambda x: (x-1)%10, **kwargs)


if __name__=='__main__':

    cifar_data      = CIFAR(cifar_root)

    train_args      = AttrDict(adv_train=False, adv_eval=False, epochs=200, lr=0.1,\
                            weight_decay=5e-4, momentum=0.9, step_lr=100, step_lr_gamma=0.1,\
                            out_dir=save_dir, log_iters=50, mixed_precision=False,\
                            save_ckpt_iters=-1, custom_lr_multiplier=None)

    model, _        = make_and_restore_model(arch='resnet50', dataset=cifar_data, resume_path=None)

    train_loader, test_loader = cifar_data.make_loaders(workers=8, batch_size=128, data_aug=True)

    model.train()
    model           = train_model(train_args, model, (train_loader, test_loader))

    ddet_data       = CIFAR(os.path.join(release_data_root, 'ddet_CIFAR'), custom_class=ReleaseDatasets)
    shifted_ddet    = CIFAR(os.path.join(release_data_root, 'ddet_CIFAR'), custom_class=ShiftedDDet)
    drand_data      = CIFAR(os.path.join(release_data_root, 'drand_CIFAR'), custom_class=ReleaseDatasets)

    eval_args       = AttrDict(adv_train=False, adv_eval=False)

    ddet_loader, _      = ddet_data.   make_loaders(workers=8, batch_size=128, data_aug=False)
    shifted_loader, _   = shifted_ddet.make_loaders(workers=8, batch_size=128, data_aug=False)
    drand_loader, _     = drand_data.  make_loaders(workers=8, batch_size=128, data_aug=False)

    if hasattr(model, "module"):
        model = model.module

    model.eval()

    print("On CIFAR test data : ")
    eval_model(eval_args, model, test_loader, None) # 92.9%

    print("On ddet dataset : ")
    eval_model(eval_args, model, ddet_loader, None) # 38.0%

    print("On Shifted ddet dataset :")
    eval_model(eval_args, model, shifted_loader, None) # 56.8%

    print("On drand dataset :")
    eval_model(eval_args, model, drand_loader, None) # 42.3%
```

**Comment by andrewilyas** [comment]
Hi, thanks for the issue! Maybe the problem is in the conversion to a PIL image and back (which usually adds some noise that could ruin the adversariality). This code gets 100% accuracy:

```python
from robustness import model_utils, datasets
from robustness.tools import folder
import os
from torch.utils.data import DataLoader
import torch as ch

data_path = "release_datasets/ddet_CIFAR/"

train_data = ch.cat(ch.load(os.path.join(data_path, f"CIFAR_ims")))
train_labels = ch.cat(ch.load(os.path.join(data_path, f"CIFAR_lab")))
train_set = folder.TensorDataset(train_data, train_labels)

tl = DataLoader(train_set, batch_size=64)

ds = datasets.CIFAR("/tmp")
m, _ = model_utils.make_and_restore_model(arch="resnet50", dataset=ds,
    resume_path="/data/theory/robustopt/robust_models/cifar_nat/checkpoint.pt.best")
m = ch.nn.DataParallel(m)
m.eval()

for im, y in tl:
    print(m(im.cuda())[0].argmax(1).cpu().eq(y).float().mean()) # prints 1.0 over and over
```

**Comment by vikramnitin9** [comment]
Thanks for the prompt reply! Much appreciated.

I tried removing the PIL conversion (code snippet below), but this had no substantial effect on the accuracy unfortunately. It's still 38.2% on ddet and 42.4% on drand.

```python
class ReleaseDatasets(torch.utils.data.Dataset):

    def __init__(self, root='data/release_datasets/ddet_CIFAR', transform=lambda x:x, target_transform=lambda x:x, **kwargs):

        self.data   = torch.cat(torch.load(os.path.join(root, 'CIFAR_ims')))
        self.labels = torch.cat(torch.load(os.path.join(root, 'CIFAR_lab')))

        # self.transform          = transforms.Compose([transforms.ToPILImage(), transform])
        self.transform          = lambda x: x
        self.target_transform   = target_transform

        super(ReleaseDatasets, self).__init__()
```

In the above example, you loaded a pre-trained network and evaluated it on ddet and drand. If it's not too much trouble, could you share some code that first trains the network and then evaluates it at the end to produce 100% accuracy on ddet and drand?

**Comment by andrewilyas** [comment]
That's correct! Sorry I didn't pay attention to your original code and thought that you were using our released (robustness library) CIFAR model.

I think the numbers you are getting actually track pretty well with adversarial transferability between two randomly trained CIFAR models at the small epsilon value we're using, here's a code snippet illustrating this:

```python
from robustness import model_utils, datasets
from robustness.tools import folder
import os
from torch.utils.data import DataLoader
import torch as ch

data_path = "release_datasets/ddet_CIFAR/"

train_data = ch.cat(ch.load(os.path.join(data_path, f"CIFAR_ims")))
train_labels = ch.cat(ch.load(os.path.join(data_path, f"CIFAR_lab")))
train_set = folder.TensorDataset(train_data, train_labels)

tl = DataLoader(train_set, batch_size=64)

ds = datasets.CIFAR("/tmp")
m, _ = model_utils.make_and_restore_model(arch="resnet50", dataset=ds,
    resume_path="/data/theory/robustopt/robust_models/cifar_nat_1/checkpoint.pt.best")
m2, _ = model_utils.make_and_restore_model(arch="resnet50", dataset=ds,
    resume_path="/data/theory/robustopt/robust_models/cifar_nat_2/checkpoint.pt.best")

m = ch.nn.DataParallel(m)
m.eval()
m2 = ch.nn.DataParallel(m2)
m2.eval()

for im, y in tl:
    new_y = (y + 1) % 10
    _, xadv = m(im.cuda(), new_y, targeted=True, make_adv=True, eps=0.5,
                constraint='2', iterations=100, step_size=0.1)
    print(m2(xadv)[0].argmax(1).cpu().eq(new_y).float().mean()) # prints numbers ~0.43
```
So I don't think there was actually anything wrong with your code!

**Comment by vikramnitin9** [comment]
Thanks Andrew, that's an interesting point. We're able to find adversarial examples that reduce the accuracy of a particular model to (close to) 0%, but the same adversarial examples transfer with an accuracy of only 43% (targeted). How would one explain this using the terminology of robust and non-robust features?

Does this mean that certain non-robust features are "more general" in some sense? As in, >>43% of the accuracy of a model is because of well-generalizing non-robust features, but different models learn different subsets of these features, and only certain of these features are learned by all models trained on the data.

Another explanation is that the remaining accuracy (92-43%) is due to _robust_ features at the epsilon = 0.5 level. The fact that we're able to reduce a single model's accuracy to 0% at the epsilon=0.5 level could be because of adversarial examples that genuinely are bugs, i.e., that don't exploit well-generalizing features in the data distribution (like Preetum showed).

Which of these two explanations do you think is correct? Or is there another possible explanation that I've overlooked?

**Comment by vikramnitin9** [comment]
Also, in Appendix D9 of your paper, you show targeted transfer success rate of adversarial examples from a trained Resnet50 to other models, and the targeted transfer success rate from Resnet50 to Resnet50 is 70+%. Is that at a different epsilon level? Or is there a large variation with different choices of hyperparameters?

I'm happy to close this issue and continue this discussion on email if you'd prefer that.

**Comment by andrewilyas** [comment]
Yes you're right, I believe the transfer experiments were done with eps=1.0 for CIFAR. This was not done for any particular reason (just different people running the code I think) but it should be noted in our paper. I'll put it in the next update). Hopefully this answers your question and thanks for bringing this to my attention!

Just as a sanity check, I verified that transfer with eps=1.0 matches D.9 exactly.

**Comment by vikramnitin9** [comment]
Thanks for the help and clarifications! I'm closing this issue now.
